internal/pkg/etcd/etcd.go

```diff
@@ -102,12 +106,49 @@ func (c *Client) ValidateForUpgrade(ctx context.Context, config config.Provider,
 			if len(member.Name) == 0 {
 				return fmt.Errorf("etcd member %d is not started, all members must be running to perform an upgrade", member.ID)
 			}
+
+			if err = validateMemberHealth(ctx, member.GetClientURLs()); err != nil {
+				return fmt.Errorf("etcd member %d is not healthy; all members must be health to perform an upgrade: %w", member.ID, err)
+			}
+		}
+
+		// Member count must not be an even number, lest the uprgade cause a loss of quorum.
+		if len(resp.Members)%2 == 0 {
```

Comment:
this is totally nitpicking, but: if len(members) = 5, quorum is 3, so upgrading even a 4-node cluster (which is even) is fine, as it leaves etcd with 2 nodes for quorum 

so not sure if there's any case except for `len(members) == 2`

Reply:
Yeah, it's true, but I'd _also_ say that in a properly-configured cluster, it means that you are already down at least one master, since the total count should be odd, and the idea here is to only upgrade one etcd node at a time so that we minimize redundancy level changes.

That said, there's always the possibility that someone sets up a _bad_ cluster which has an even number of etcd nodes in its quiescent state, which would then leave the cluster unable to upgrade at all.  One could say that is a _good_ thing, since it enforces a non-even-numbered rule... but that also presents a possibility of unexpected behaviour.

I'm kind of ambivalent about it, either way.... so other opinions?

Reply:
I don't care much about that, but on other hand we allow clusters to have _any_ number of control plane nodes, and the fact that we don't allow upgrades for clusters with even number of nodes might come as a surprise

Reply:
Yeah... we don't really have any control at all on the number of master/etcd nodes we have.  The surprise element is the only thing which concerns me, and I really don't mind enforcing best practices here, so long as the failure cause is visible.